src/libcore/slice/mod.rs

```diff
@@ -1402,6 +1402,172 @@ impl<T> [T] {
         sort::quicksort(self, |a, b| f(a).lt(&f(b)));
     }
 
+    /// Moves all consecutive repeated elements to the end of the slice.
+    /// Returns two slices, the first contains the deduplicated elements,
+    /// the second contains all the duplicates unordered.
+    ///
+    /// If the slice is sorted, the first slice contains no duplicates.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(slice_retain_dedup)]
+    ///
+    /// let mut slice = [1, 2, 2, 3, 3, 2];
+    ///
+    /// let (dedup, duplicates) = slice.retain_dedup();
+    ///
+    /// assert_eq!(dedup, [1, 2, 3, 2]);
+    /// assert_eq!(duplicates, [3, 2]);
+    /// ```
+    #[unstable(feature = "slice_retain_dedup", issue = "0")]
+    #[inline]
+    pub fn retain_dedup(&mut self) -> (&mut [T], &mut [T])
+        where T: PartialEq
+    {
+        self.retain_dedup_by(|a, b| a == b)
+    }
+
+    /// Moves all but the first of consecutive elements to the end of the slice satisfying
+    /// a given equality relation. Returns two slices, the first contains the deduplicated elements,
+    /// the second contains all the duplicates unordered.
+    ///
+    /// The `same_bucket` function is passed references to two elements from the slice, and
+    /// returns `true` if the elements compare equal, or `false` if they do not. The elements are
+    /// passed in opposite order from their order in the slice, so if `same_bucket(a, b)` returns
+    /// `true`, `a` is removed.
+    ///
+    /// If the slice is sorted, the first slice contains no duplicates.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(slice_retain_dedup)]
+    ///
+    /// let mut slice = ["foo", "Foo", "bar", "Bar", "baz", "bar"];
```

Comment:
I would extend this array to:
```rust
let mut slice = ["foo", "Foo", "FOO", "bar", "Bar", "baz", "bar"];
```
to make the algorithm more clear wrt. what happens with the order in `duplicates`.

Reply:
Extending this array will not help to understand what really happen to the `duplicates`. The resulting `duplicates` array will be `["Bar", "FOO", "Foo"]` which could be interpreted as a single push front stack.

I would prefer to use this pattern instead: `["foo", "Foo", "BAZ", "Bar", "bar", "baz", "BAZ"]`, the `duplicates` part would be `["bar", "Foo", "BAZ"]`, which didn't miss-inform the user on the resulting non-order.

Reply:
I will change the other examples to be explicit, the user must not think that the duplicates are ordered in any way.